File: hora/horoscope/chart/arudhas.py
from hora import const, utils
from hora.horoscope.chart.house import *
def bhava_arudhas_from_planet_positions(planet_positions):
    """
        gives Bhava Arudhas for each house from the chart
        @param planet_positions: Planet Positions in the format: \
        [ [planet,[rasi,longitude]], [[,]].., [[,]]]
        @return bhava arudhas of houses. first element is for the first house from lagna and so on
    """
    """
        TODO: Check if A11 is calculated correct for tajaka charts?
    """
    h_to_p = utils.get_house_planet_list_from_planet_positions(planet_positions)
    #print('h_to_p',h_to_p)
    p_to_h = utils.get_planet_to_house_dict_from_chart(h_to_p)
    #print('p_to_h',p_to_h)
    asc_house = p_to_h[const._ascendant_symbol]
    houses = [(h+asc_house)%12 for h in range(12)]
    bhava_arudhas_of_houses =[]
    for i,h in enumerate(houses):
        # get house lord. For Aq/Sc get stronger lord
        #print('house',i,'in rasi',rasi_names_en[h])
        lord_of_the_house = house_owner_from_planet_positions(planet_positions, h, check_during_dhasa=False)
        house_of_the_lord = p_to_h[lord_of_the_house]
        #print('house',i,'lord_of_the_house',lord_of_the_house,'house_of_the_lord',house_of_the_lord)
        #print(planet_list[lord_of_the_house],rasi_names_en[house_of_the_lord])
        signs_between_house_and_lord = (house_of_the_lord+1+12-h)%12
        #print('signs_between_house_and_lord',h,house_of_the_lord,signs_between_house_and_lord)
        bhava_arudha_of_house = (house_of_the_lord+signs_between_house_and_lord-1)%12
        signs_from_the_house = ((bhava_arudha_of_house+1+12-h)%12)
        #print('signs_from_the_house',signs_from_the_house)
        if signs_from_the_house in [1,7]:
            #print('in [1,7] from the original house')
            bhava_arudha_of_house = (bhava_arudha_of_house+10-1)%12
        #print('bhava arudha:','A'+str(i+1),'is',bhava_arudha_of_house,rasi_names_en[bhava_arudha_of_house])
        bhava_arudhas_of_houses.append(bhava_arudha_of_house)
    return bhava_arudhas_of_houses
def bhava_arudhas(chart):
    """
        gives Bhava Arudhas for each house from the chart
        @param chart: Enter chart information in the following format. For each house from Aries planet numbers separated by /
            ['0/1','2','','','3/4/5','','','6','L/7','','8','']
        @return bhava arudhas of houses. first element is for the first house from lagna and so on
    """
    h_to_p = chart[:]
    #print('h_to_p',h_to_p)
    p_to_h = utils.get_planet_to_house_dict_from_chart(h_to_p)
    #print('p_to_h',p_to_h)
    asc_house = p_to_h[const._ascendant_symbol]
    houses = [(h+asc_house)%12 for h in range(12)]
    bhava_arudhas_of_houses =[]
    for i,h in enumerate(houses):
        # get house lord. For Aq/Sc get stronger lord
        #print('house',i,'in rasi',rasi_names_en[h])
        lord_of_the_house = house_owner(h_to_p, h) # V2.3.1
        house_of_the_lord = p_to_h[lord_of_the_house]
        #print('house',i,'lord_of_the_house',lord_of_the_house,'house_of_the_lord',house_of_the_lord)
        #print(planet_list[lord_of_the_house],rasi_names_en[house_of_the_lord])
        signs_between_house_and_lord = (house_of_the_lord+1+12-h)%12
        #print('signs_between_house_and_lord',h,house_of_the_lord,signs_between_house_and_lord)
        bhava_arudha_of_house = (house_of_the_lord+signs_between_house_and_lord-1)%12
        signs_from_the_house = ((bhava_arudha_of_house+1+12-h)%12)
        #print('signs_from_the_house',signs_from_the_house)
        if signs_from_the_house in [1,7]:
            #print('in [1,7] from the original house')
            bhava_arudha_of_house = (bhava_arudha_of_house+10-1)%12
        #print('bhava arudha:','A'+str(i+1),'is',bhava_arudha_of_house,rasi_names_en[bhava_arudha_of_house])
        bhava_arudhas_of_houses.append(bhava_arudha_of_house)
    return bhava_arudhas_of_houses
def graha_arudhas_from_planet_positions(planet_positions):
    """
        gives Graha Arudhas for each planet from the planet positions
        @param planet_positions: Planet Positions in the format: \
        [ [planet,[rasi,longitude]], [[,]].., [[,]]]
        @return graha arudhas of planet. first element is for Sun, last element is for Ketu
    """
    h_to_p = utils.get_house_planet_list_from_planet_positions(planet_positions)
    #print('h_to_p',h_to_p)
    p_to_h = utils.get_planet_to_house_dict_from_chart(h_to_p)
    #print('p_to_h',p_to_h)
    asc_house = p_to_h[const._ascendant_symbol]
    graha_arudhas_of_planets = []
    for p,planet in enumerate(planet_list):
        house_of_the_planet = p_to_h[p]
        #print('house_of_the_planet',house_of_the_planet,rasi_names_en[house_of_the_planet])
        sign_owned_by_planet = const.house_lords_dict[p]
        #print('sign_owned_by_planet',sign_owned_by_planet)
        if len(sign_owned_by_planet)>1:
            #print('stronger of',sign_owned_by_planet[0],sign_owned_by_planet[1])
            sign_owned_by_planet = stronger_rasi_from_planet_positions(planet_positions,sign_owned_by_planet[0],sign_owned_by_planet[1])
            #print('stronger raasi', sign_owned_by_planet)
        else:
            sign_owned_by_planet = sign_owned_by_planet[0]
        #print(planet,rasi_names_en[sign_owned_by_planet])
        count_to_strong = (sign_owned_by_planet+1+12-house_of_the_planet)%12
        #print('count_to_strong',count_to_strong)
        count_to_arudha = (house_of_the_planet+2*(count_to_strong-1))%12
        #print('count again',count_to_arudha,rasi_names_en[count_to_arudha])
        count_from_house = (house_of_the_planet+12-count_to_arudha)%12
        #print('count_from_house',count_from_house)
        if count_from_house in [0,6]:
            count_to_arudha = (count_to_arudha+9)%12
            #print('count in [1,7] from house',count_to_arudha)
        graha_padha_of_planet = count_to_arudha
        #print(planet_list[p],rasi_names_en[graha_padha_of_planet],'\n')
        graha_arudhas_of_planets.append(graha_padha_of_planet)
    return graha_arudhas_of_planets
def graha_arudhas(chart):
    """
        gives Graha Arudhas for each planet from the chart
        @param chart: Enter chart information in the following format. For each house from Aries planet numbers separated by /
            ['0/1','2','','','3/4/5','','','6','L/7','','8','']
        @return graha arudhas of planet. first element is for Sun, last element is for Ketu
    """
    h_to_p = chart[:]
    #print('h_to_p',h_to_p)
    p_to_h = utils.get_planet_to_house_dict_from_chart(h_to_p)
    #print('p_to_h',p_to_h)
    #print('house_lords_dict',house_lords_dict)
    asc_house = p_to_h[const._ascendant_symbol]
    graha_arudhas_of_planets = []
    for p,planet in enumerate(planet_list):
        house_of_the_planet = p_to_h[p]
        #print('house_of_the_planet',house_of_the_planet,rasi_names_en[house_of_the_planet])
        sign_owned_by_planet = const.house_lords_dict[p]
        #print('sign_owned_by_planet',sign_owned_by_planet)
        if len(sign_owned_by_planet)>1:
            #print('stronger of',sign_owned_by_planet[0],sign_owned_by_planet[1])
            sign_owned_by_planet = stronger_rasi(h_to_p,sign_owned_by_planet[0],sign_owned_by_planet[1])
            #print('stronger raasi', sign_owned_by_planet)
        else:
            sign_owned_by_planet = sign_owned_by_planet[0]
        #print(planet,rasi_names_en[sign_owned_by_planet])
        count_to_strong = (sign_owned_by_planet+1+12-house_of_the_planet)%12
        #print('count_to_strong',count_to_strong)
        count_to_arudha = (house_of_the_planet+2*(count_to_strong-1))%12
        #print('count again',count_to_arudha,rasi_names_en[count_to_arudha])
        count_from_house = (house_of_the_planet+12-count_to_arudha)%12
        #print('count_from_house',count_from_house)
        if count_from_house in [0,6]:
            count_to_arudha = (count_to_arudha+9)%12
            #print('count in [1,7] from house',count_to_arudha)
        graha_padha_of_planet = count_to_arudha
        #print(planet_list[p],rasi_names_en[graha_padha_of_planet],'\n')
        graha_arudhas_of_planets.append(graha_padha_of_planet)
    return graha_arudhas_of_planets
    
if __name__ == "__main__":
    from hora.tests.pvr_tests import test_example
    from hora.horoscope.chart import house, charts
    def chapter_9_tests():
        chapter = 'Chapter 9.2 Bhava/Graha Arudhas '
        chart_1 = ['4/2/6','','1','7','','L','','','','8','','3/0/5']
        chart_1_dob = (2000,4,9)
        chart_1_tob = (17,55,0)
        chart_1_place = drik.Place('unknwon',42+30/60,-71-12/60,-4.0)
        chart_1_dcf = 1
        chart_2 = ['6','5','','7/8','','','','','3/L','4','1/0/2','']
        chart_2_dob = chart_1_dob
        chart_2_tob = chart_1_tob
        chart_2_place = drik.Place('unknwon',42+30/60,-71-12/60,-5.0)
        chart_2_dcf = 16
        def bhava_arudha_tests_1():
            exercise = 'Example 29 / Chart 1 Bhava Arudha'
            jd_at_dob = utils.julian_day_number(chart_1_dob, chart_1_tob)
            planet_positions = charts.divisional_chart(jd_at_dob, chart_1_place, divisional_chart_factor=chart_1_dcf)
            asc_house = planet_positions[0][1][0]
            ba = bhava_arudhas_from_planet_positions(planet_positions)#bhava_arudhas(chart_1)
            expected_result = [2, 4, 5, 4, 0, 2, 1, 9, 9, 5, 1, 6]
            #Ans A1/AL:Ge, A2:Le, A3:Vi, A4:Le, A5:Ar, A6:Ge, A7:Ta, A8:Cp, A9:Cp, A10:Vi, A11: Ta, A12:Li
            houses = [(h + asc_house) % 12 for h in range(12)] 
            for i, h in enumerate(houses):
                test_example(chapter+exercise,house.rasi_names_en[expected_result[i]],house.rasi_names_en[ba[i]],'A' + str(i + 1))
        def bhava_arudha_tests_2():
            exercise = 'Exercise 12 / Chart 2 Bhava Arudha'
            jd_at_dob = utils.julian_day_number(chart_2_dob, chart_2_tob)
            planet_positions = charts.divisional_chart(jd_at_dob, chart_2_place, divisional_chart_factor=chart_2_dcf)
            asc_house = planet_positions[0][1][0]
            ba = bhava_arudhas_from_planet_positions(planet_positions)#ba = bhava_arudhas(chart_2)
            expected_result = [10,0,8,7,8,10,11,5,1,8,8,10]
            houses = [(h + asc_house) % 12 for h in range(12)] 
            for i, h in enumerate(houses):
                test_example(chapter+exercise,house.rasi_names_en[expected_result[i]],house.rasi_names_en[ba[i]],'A' + str(i + 1))
        def graha_arudha_tests_1():
            exercise = 'Example 29 / Chart 1 Graha Arudha'
            jd_at_dob = utils.julian_day_number(chart_1_dob, chart_1_tob)
            planet_positions = charts.divisional_chart(jd_at_dob, chart_1_place, divisional_chart_factor=chart_1_dcf)
            asc_house = planet_positions[0][1][0]
            ba = graha_arudhas_from_planet_positions(planet_positions)#graha_arudhas(chart_1)
            expected_result = [9,4,9,2,10,1,3,5,5]
            for p in range(9):
                test_example(chapter+exercise,house.rasi_names_en[expected_result[p]],house.rasi_names_en[ba[p]],'contains',house.planet_list[p],"Graha Pada")
        def graha_arudha_tests_2():
            exercise = 'Exercise 13 / Chart 2 Graha Arudha'
            jd_at_dob = utils.julian_day_number(chart_2_dob, chart_2_tob)
            planet_positions = charts.divisional_chart(jd_at_dob, chart_2_place, divisional_chart_factor=chart_2_dcf)
            asc_house = planet_positions[0][1][0]
            ba = graha_arudhas_from_planet_positions(planet_positions)#ba = graha_arudhas(chart_2)
            expected_result = [7,8,2,11,7,10,8,5,11]
            for p in range(9):
                test_example(chapter+exercise,house.rasi_names_en[expected_result[p]],house.rasi_names_en[ba[p]],'contains',house.planet_list[p],"Graha Pada")
        bhava_arudha_tests_1()
        bhava_arudha_tests_2()
        graha_arudha_tests_1()
        graha_arudha_tests_2()
    chapter_9_tests()

File: pyjhora-web/src/core/horoscope/arudhas.ts
/**
 * Arudha Pada Calculations
 * Ported from PyJHora horoscope/chart/arudhas.py
 *
 * Arudhas are special lagnas (ascendants) that represent the image or
 * manifestation of houses and planets. They are extensively used in
 * Jaimini astrology.
 *
 * Types of Arudhas:
 * - Bhava Arudhas (A1-A12): Arudhas of the 12 houses
 *   - A1 is also called Arudha Lagna (AL)
 *   - A12 is also called Upa Lagna (UL)
 * - Surya Arudhas (S1-S12): Bhava Arudhas calculated from Sun's position
 * - Chandra Arudhas (M1-M12): Bhava Arudhas calculated from Moon's position
 * - Graha Arudhas: Arudhas of planets (Pada of each planet)
 */

import {
  PLANET_SIGNS_OWNED,
  PP_COUNT_UPTO_KETU,
  PLANETS_UPTO_KETU,
  SUN,
  MOON,
} from '../constants';

import { countRasis } from '../utils/angle';

import {
  getHouseOwnerFromPlanetPositions,
  getPlanetToHouseDict,
  getHouseToPlanetList,
  getStrongerRasi,
} from './house';

/**
 * Planet position format used for Arudha calculations
 */
export interface ArudhaPlanetPosition {
  planet: number;
  rasi: number;
  longitude: number;
}

/**
 * Calculate Bhava Arudhas for each house from planet positions
 *
 * Bhava Arudhas are calculated as follows:
 * 1. Find the lord of the house
 * 2. Count signs from house to lord's position
 * 3. Count same number from lord's position
 * 4. If result falls in 1st or 7th from original house, add 10 signs
 *
 * @param planetPositions - Array of planet positions (must include at least Lagna to Ketu)
 * @param arudhaBase - Base planet for calculation:
 *   - 0: Lagna (default) - returns A1, A2, ... A12
 *   - 1: Sun - returns Surya Arudhas S1, S2, ... S12
 *   - 2: Moon - returns Chandra Arudhas M1, M2, ... M12
 *   - 3-9: Other planets (Mars to Ketu)
 * @returns Array of 12 rasi indices representing Bhava Arudhas for houses 1-12 from base
 */
export function bhavaArudhasFromPlanetPositions(
  planetPositions: ArudhaPlanetPosition[],
  arudhaBase: number = 0
): number[] {
  // Restrict to planets up to Ketu (exclude Uranus/Neptune/Pluto)
  // This is crucial for correct results as per PyJHora V3.6.4
  const positions = planetPositions.slice(0, PP_COUNT_UPTO_KETU);

  // Build helper dictionaries
  const houseToPlanetList = getHouseToPlanetList(positions);
  const planetToHouse = getPlanetToHouseDict(positions);

  // Get the base house (rasi where base planet is located)
  const baseHouse = positions[arudhaBase]?.rasi ?? 0;

  // Calculate houses from the base
  const houses = Array.from({ length: 12 }, (_, h) => (h + baseHouse) % 12);

  const bhavaArudhasOfHouses: number[] = [];

  for (const h of houses) {
    // Step 1: Find lord of the house
    const lordOfTheHouse = getHouseOwnerFromPlanetPositions(positions, h, false);

    // Step 2: Find house where lord is placed
    const houseOfTheLord = planetToHouse[lordOfTheHouse] ?? 0;

    // Step 3: Count signs from house to lord's position (inclusive)
    const signsBetweenHouseAndLord = countRasis(h, houseOfTheLord);

    // Step 4: Calculate Bhava Arudha position
    // Count same number of signs from lord's position
    let bhavaArudhaOfHouse = (houseOfTheLord + signsBetweenHouseAndLord - 1) % 12;

    // Step 5: Check if Arudha falls in 1st or 7th from original house
    const signsFromTheHouse = countRasis(h, bhavaArudhaOfHouse);

    // Exception: If Arudha is in 1st or 7th house from original, add 10 signs
    if (signsFromTheHouse === 1 || signsFromTheHouse === 7) {
      bhavaArudhaOfHouse = (bhavaArudhaOfHouse + 10 - 1) % 12;
    }

    bhavaArudhasOfHouses.push(bhavaArudhaOfHouse);
  }

  return bhavaArudhasOfHouses;
}

/**
 * Calculate Surya (Sun) Arudhas - Bhava Arudhas calculated from Sun's position
 * @param planetPositions - Array of planet positions
 * @returns Array of 12 rasi indices for Surya Arudhas S1-S12
 */
export function suryaArudhasFromPlanetPositions(
  planetPositions: ArudhaPlanetPosition[]
): number[] {
  return bhavaArudhasFromPlanetPositions(planetPositions, SUN + 1); // +1 because index 0 is Lagna
}

/**
 * Calculate Chandra (Moon) Arudhas - Bhava Arudhas calculated from Moon's position
 * @param planetPositions - Array of planet positions
 * @returns Array of 12 rasi indices for Chandra Arudhas M1-M12
 */
export function chandraArudhasFromPlanetPositions(
  planetPositions: ArudhaPlanetPosition[]
): number[] {
  return bhavaArudhasFromPlanetPositions(planetPositions, MOON + 1); // +1 because index 0 is Lagna
}

/**
 * Calculate Graha Arudhas (Pada) for each planet
 *
 * Graha Arudha (Planetary Pada) calculation:
 * 1. Find the house where planet is placed
 * 2. Find the stronger sign owned by the planet
 * 3. Count from planet's house to the owned sign
 * 4. Count same distance from planet's house
 * 5. If result is 1st or 7th from planet's house, add 10 signs
 *
 * @param planetPositions - Array of planet positions
 * @returns Array of rasi indices for Graha Arudhas
 *          Index 0: Lagna Pada
 *          Index 1-9: Sun Pada, Moon Pada, Mars Pada, etc.
 */
export function grahaArudhasFromPlanetPositions(
  planetPositions: ArudhaPlanetPosition[]
): number[] {
  const positions = planetPositions.slice(0, PP_COUNT_UPTO_KETU);
  const planetToHouse = getPlanetToHouseDict(positions);

  // First element is Lagna's position (Lagna Pada = Lagna's house)
  const grahaArudhasOfPlanets: number[] = [positions[0]?.rasi ?? 0];

  // Calculate for each planet from Sun (0) to Ketu (8)
  for (let p = 0; p < PLANETS_UPTO_KETU; p++) {
    const houseOfThePlanet = planetToHouse[p] ?? 0;

    // Get signs owned by this planet
    let signOwnedByPlanet = PLANET_SIGNS_OWNED[p];

    if (!signOwnedByPlanet || signOwnedByPlanet.length === 0) {
      // Fallback (should not happen with proper data)
      grahaArudhasOfPlanets.push(houseOfThePlanet);
      continue;
    }

    // If planet owns two signs, find the stronger one
    let strongerSign: number;
    if (signOwnedByPlanet.length > 1) {
      strongerSign = getStrongerRasi(
        positions,
        signOwnedByPlanet[0],
        signOwnedByPlanet[1]
      );
    } else {
      strongerSign = signOwnedByPlanet[0];
    }

    // Count from planet's house to its stronger owned sign
    // Formula: (sign + 1 + 12 - house) % 12 in Python
    // This gives 0-based distance, we need 1-based for countRasis
    const countToStrong = ((strongerSign + 1 + 12 - houseOfThePlanet) % 12);

    // Calculate Arudha position: house + 2*(count - 1)
    let countToArudha = (houseOfThePlanet + 2 * (countToStrong - 1)) % 12;

    // Check if Arudha is in 1st or 7th from planet's house
    const countFromHouse = (houseOfThePlanet + 12 - countToArudha) % 12;

    // Exception: If count is 0 or 6 (same house or 7th house), add 9 signs
    if (countFromHouse === 0 || countFromHouse === 6) {
      countToArudha = (countToArudha + 9) % 12;
    }

    const grahaPadhaOfPlanet = countToArudha;
    grahaArudhasOfPlanets.push(grahaPadhaOfPlanet);
  }

  return grahaArudhasOfPlanets;
}

/**
 * Get Arudha Lagna (A1) - the most commonly used Arudha
 * @param planetPositions - Array of planet positions
 * @returns Rasi index of Arudha Lagna
 */
export function getArudhaLagna(planetPositions: ArudhaPlanetPosition[]): number {
  const bhavaArudhas = bhavaArudhasFromPlanetPositions(planetPositions, 0);
  return bhavaArudhas[0]; // A1 is the first element
}

/**
 * Get Upa Lagna (A12) - Arudha of the 12th house
 * @param planetPositions - Array of planet positions
 * @returns Rasi index of Upa Lagna
 */
export function getUpaLagna(planetPositions: ArudhaPlanetPosition[]): number {
  const bhavaArudhas = bhavaArudhasFromPlanetPositions(planetPositions, 0);
  return bhavaArudhas[11]; // A12 is the 12th element (index 11)
}

/**
 * Format Bhava Arudhas as a chart array (for display)
 *
 * @param bhavaArudhas - Array of 12 Bhava Arudha positions
 * @param prefix - Prefix for labels (default 'A' for Bhava Arudhas)
 * @returns Array of 12 strings, one for each rasi, containing Arudha labels
 *          e.g., ['A1/A3', '', 'A2', ...] for Aries, Taurus, etc.
 */
export function formatBhavaArudhasAsChart(
  bhavaArudhas: number[],
  prefix: string = 'A'
): string[] {
  const chart: string[] = Array(12).fill('');

  bhavaArudhas.forEach((rasi, index) => {
    const label = `${prefix}${index + 1}`;
    if (chart[rasi]) {
      chart[rasi] += `/${label}`;
    } else {
      chart[rasi] = label;
    }
  });

  return chart;
}

/**
 * Format Graha Arudhas as a chart array (for display)
 *
 * @param grahaArudhas - Array of Graha Arudha positions (Lagna + 9 planets)
 * @returns Array of 12 strings, one for each rasi, containing planet labels
 *          e.g., ['L/0', '3', '', ...] where L=Lagna, 0=Sun, 3=Mercury
 */
export function formatGrahaArudhasAsChart(grahaArudhas: number[]): string[] {
  const chart: string[] = Array(12).fill('');

  grahaArudhas.forEach((rasi, index) => {
    // Index 0 is Lagna, 1-9 are planets (Sun=0 to Ketu=8 in planet numbering)
    const label = index === 0 ? 'L' : String(index - 1);
    if (chart[rasi]) {
      chart[rasi] += `/${label}`;
    } else {
      chart[rasi] = label;
    }
  });

  return chart;
}
